### DISCUSSION TITLE
exploring-the-dotnet-8-preview-introducing-the-identity-api-endpoints/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#388] [Announcements]
# exploring-the-dotnet-8-preview-introducing-the-identity-api-endpoints/

In this post I introduce the new Identity APIs, show how they differ from the existing Identity support, and show how to use them.

https://andrewlock.net/exploring-the-dotnet-8-preview-introducing-the-identity-api-endpoints/

**Comment by AmyJeanes** [comment]
You have a minor typo `MappIdentityApi` :)

**Comment by solo12zw74** [comment]
Hello Andrew. Thank you for these detailed reviews.
I am curious about the question about Swagger, I see you're using it in your sources, but as long as I can't found the full sources - I can't find the library you're using for work with OpenAPI/Swagger. My question is - what do you personally prefer or can recommend to use: NSwag or Swahbuckle toolset to work with OPenAPI specs (doc generation, client generation)?
Thank you.

**Reply by andrewlock** [reply]
The example shown there is using Swashbuckle, as that's what the default ASP.NET Core templates.

From my experience, I've generally found Swashbuckle for document generation and NSwag for client generation to be preferable.

That said, I've personally found quite a lot of friction with client generation, and ended up manually creating clients in my previous role, so your mileage may vary 🤷‍♂️

**Reply by solo12zw74** [reply]
Thank for replying.
Yes, we ara currently are looking the easier and frictionless way of generating client. And currently we have Swashbuckle for doc gen. But client gen is made with different version of Java tool (swagger-generator, then openapi-generator) whici stored in a repo. And I am considering that approach not very eefective, while Nswag (if I know that correct) is able to make both things - openapi doc gen and client gen. Just sharing. Thank you for your time.

**Comment by akhanalcs** [comment]
Great article. Can you please expand on this series because Identity in .NET can sure use some more blog posts and docs?
Can you also please create a new blog post on how to use Identity with some Oauth provider to call protected APIs, machine-machine communication and so on?
Thanks!

**Comment by oliverw** [comment]
Is there support for acquiring an access token for your SPA/Blazor WASM client when your backend relies solely on logging in users using external login providers like Microsoft, Google etc?

**Reply by andrewlock** [reply]
Nope, that's explicitly not supported

**Reply by oliverw** [reply]
Any idea why?

**Comment by AxharKhan** [comment]
I am using API Controller for the endpoints, but the Authorize attribute is not working with the configuration you provided I am trying swagger and it automatically returns 200 instead of unauthorized.

**Reply by AxharKhan** [reply]
Nevermind, I signed in to test and I didnt think swagger would store the authentication cookie, I cleared the cookies and it is working fine now.

**Comment by Wayne-Mather** [comment]
Hi Andrew,

Do you know how to disable a certain endpoint? Say you don't want register to be public, how do you disable them without rewriting the add api endpoint extension?

**Reply by andrewlock** [reply]
Nope - I suggest you read this post too https://andrewlock.net/should-you-use-the-dotnet-8-identity-api-endpoints/

**Comment by mariohik** [comment]
Hello Andrew! 
Is it possible to let only Admin users access the register endpoint?

**Comment by Brainforce1** [comment]
When having a custom IdentityUser (in .NET 8), with some additional fields, the Register endpoint seems to fail when called, stating that those additional field(s) cannot be NULL. 
Even though I do provide the values for them, it seems that they're not being used or detected.
Do you know what might be the cause of this? I cannot find any good resources about this, and Microsoft's otherwise excellent 'learn' site is also not saying anything about this issue.

**Comment by moosaam** [comment]
I want to add role here, based upon certain user role, I want to restrict the access. How we can achieve this. In Identity api, I can't find role ,user role end point. But it's exists in generated tables in sql server.

**Comment by tomerfir** [comment]
I was trying to use Identity now myself, and I noticed that in your register example you set `username` and `email` to the same value:

```C#
### Register a new user
POST http://localhost:5117/account/register
Content-Type: application/json

{
  "username": "andrew@example.com",
  "password": "SuperSecret1!",
  "email": "andrew@example.com"
}
```
And I tested it myself and found that setting the `username` field in the register requests literally does nothing. In fact, you don't even have to send it, and it seems like in the database the user's `username` is set to the email regardless of what you write in the `username` field of the request.

Am I the only one that finds this a little absurd? I couldn't find anything about this, and I don't know whether its maybe just a bug due to me using NpgSQL, but it feels crazy to me that the `username` field is just an alternate way to access the email.

**Comment by worbarry** [comment]
what if i already have a webapp and a made a sepreate API so i can also use auth in my desktop app how do i connect the webapp to the api for auth instead of been local to the webapp?
